tensorflow/core/lib/png/png_io.cc

```diff
@@ -232,11 +232,19 @@ bool CommonInitDecode(StringPiece png_string, int desired_channels,
     CommonFreeDecode(context);
     return false;
   }
-  if (context->channels == 0) {  // Autodetect number of channels
-    context->channels = png_get_channels(context->png_ptr, context->info_ptr);
-  }
   const bool has_tRNS =
       (png_get_valid(context->png_ptr, context->info_ptr, PNG_INFO_tRNS)) != 0;
+  if (context->channels == 0) {  // Autodetect number of channels
+    if (context->color_type == PNG_COLOR_TYPE_PALETTE) {
```

Comment:
I'm not familiar with this code -- why do we need to handle PNG_COLOR_TYPE_PALETTE specially? Is it because we convert palette to rgb below?

Reply:
Certainly, you convert palette to rgb bellow.
But grayscale conversion run after that, because channels is 1 (args channels=0)

This issue depends on meaning of:
 "the number of channels in the PNG-encoded image."
https://www.tensorflow.org/api_docs/python/tf/image/decode_png

png_get_channels function return 1 if colortype is PALETTE.
http://www.libpng.org/pub/png/libpng-manual.html#section-3.6
Palette has indices channels so channels 1 only.

I think that the current behavior is unkind.

Reply:
Ah I see, thank you for the explanation. I hadn't looked at the documentation before, but now it makes sense.